File: test/jsdom/ceremonyname.test.mjs
// Naming a proceeding on this machine, at the surface.
//
// **What only this tier can see.** The Go tests prove the route stores, caps and clears a name,
// and that the listing carries it. They cannot see which of the two strings ends up as the card's
// heading, whether the recital survives being demoted, or whether the field a user types into is
// refilled from the card rather than from whatever they abandoned last time.
//
// **The recital assertion is the one worth writing carefully.** D20 makes the Intent the recital's
// only home, so a name that REPLACED it would put this panel in the business of paraphrasing what
// the parties agreed — and the failure is silent, because a card with a good name on it looks
// entirely correct. So both strings are asserted present, on the same card, in that order.
//
// One boot per file — see boot.mjs.
import test from 'node:test';
import assert from 'node:assert/strict';
import { boot } from './boot.mjs';

const ID = '7'.repeat(32);
const INTENT = 'We agree to the lease of 14 Elm Row, Edinburgh, for a term of five years';

let listing = {};
let named = null;

const h = await boot({
  routes: {
    '/api/ceremonies': () => listing,
    // `opts`, not a parsed body: the harness hands the route the fetch options, so what is
    // recorded here is what the client actually SENT.
    '/api/ceremony/name': (opts) => {
      named = JSON.parse((opts && opts.body) || '{}');
      return { ceremony: ID, name: named.name || '' };
    },
    '/api/ceremony/next': () => ({ ceremony: ID, state: 'unavailable', reason: 'no record' }),
  },
});
const { document: doc, settle } = h;

const card = (extra) => Object.assign({
  id: ID, state: 'absent', me: 'b'.repeat(64), intent: INTENT,
  reason: 'nothing has arrived yet',
}, extra || {});

async function render(data) {
  listing = data;
  named = null;
  doc.querySelector('.modetab[data-tab="collaborate"]').click();
  await settle();
}

test('an unnamed ceremony is still headed by its recital, and says it once', async () => {
  await render({ ceremonies: [card()], primary: true });
  assert.equal(doc.querySelector('.cerintent').textContent, INTENT,
    'an unnamed ceremony lost its heading — the recital is what it had, and the name is an '
    + 'addition to the card rather than a replacement for it');
  assert.equal(doc.querySelector('.cerrecital'), null,
    'the recital is printed twice on an unnamed ceremony: once as the heading and once as the '
    + 'line under it');
});

test('a named ceremony is headed by the name AND still carries the recital', async () => {
  await render({ ceremonies: [card({ name: 'The Elm Row lease' })], primary: true });
  assert.equal(doc.querySelector('.cerintent').textContent, 'The Elm Row lease',
    'the card is headed by the recital even though this machine gave it a name — which is the '
    + 'whole reason the name exists: a clause meant to be signed is not a handle to scan a list by');
  const recital = doc.querySelector('.cerrecital');
  assert.ok(recital, 'naming the ceremony HID the recital. D20 makes the Intent the recital\'s '
    + 'only home, so a name that replaces it leaves the panel paraphrasing what the parties agreed');
  assert.equal(recital.textContent, INTENT, 'the recital line does not hold the Intent');
});

test('the field opens holding the name the card carries, and Save posts it', async () => {
  await render({ ceremonies: [card({ name: 'The Elm Row lease' })], primary: true });
  const open = doc.querySelector('.cernamebtn');
  assert.match(open.textContent, /rename/i,
    'a ceremony that already has a name offers "Name this ceremony…", which reads as though the '
    + 'name were gone');
  const body = doc.querySelector('.cernamebody');
  assert.equal(body.hidden, true, 'every card in the list opens with a text field showing');
  open.click();
  await settle();
  const field = doc.querySelector('.cernameinput');
  assert.equal(body.hidden, false, 'pressing the button did not open the field');
  assert.equal(field.value, 'The Elm Row lease',
    `the field opened holding ${JSON.stringify(field.value)} — a rename that starts empty reads `
    + 'as though the name had already been cleared, and a Save on it would clear it');

  field.value = 'Elm Row';
  doc.querySelector('.cernamesave').click();
  await settle();
  assert.ok(named, 'Save posted nothing');
  assert.equal(named.ceremony, ID, `the request named ceremony ${JSON.stringify(named.ceremony)}`);
  assert.equal(named.name, 'Elm Row', `the request carried ${JSON.stringify(named.name)}`);
});

test('an emptied field is SENT, because clearing has no second door', async () => {
  await render({ ceremonies: [card({ name: 'The Elm Row lease' })], primary: true });
  doc.querySelector('.cernamebtn').click();
  await settle();
  const field = doc.querySelector('.cernameinput');
  field.value = '';
  doc.querySelector('.cernamesave').click();
  await settle();
  // The assertion is that a request was made AT ALL with an empty name: a client that skipped the
  // post on an empty field would leave the user with no way back to the agreement, and the card
  // would still be showing the name they just deleted.
  assert.ok(named, 'clearing the field and pressing Save sent nothing, so a name cannot be removed '
    + 'from the product at all — the server treats empty as "forget it" and nothing reaches it');
  assert.equal(named.name, '', `the cleared request carried ${JSON.stringify(named.name)}`);
});

test('the name says plainly that it never leaves this machine', async () => {
  await render({ ceremonies: [card()], primary: true });
  doc.querySelector('.cernamebtn').click();
  await settle();
  const note = doc.querySelector('.cernamenote').textContent;
  assert.match(note, /this machine/i,
    `the box says ${JSON.stringify(note)}. The name is local, unsigned and never on the wire — a `
    + 'user who believes the other parties see their label would write a different label');
  assert.match(note, /signs?\b/i, 'the box does not say the name is outside what anyone signs');
});

test('a non-primary Nib is not offered the control', async () => {
  await render({ ceremonies: [card()], primary: false });
  assert.equal(doc.querySelector('.cernamebtn'), null,
    'a non-primary Nib offers a control that writes to ~/nib — `mayAct` folds that together with '
    + 'the lock screen, and both are states where the user is looking rather than acting');
});

test('a finished ceremony is still identifiable by its name', async () => {
  await render({
    ceremonies: [],
    ended: [{ ceremony: ID, state: 'completed', observed_at: '2026-09-01T10:00:00Z', name: 'The Elm Row lease' }],
    primary: true,
  });
  const row = doc.querySelector('.cerended-row');
  assert.ok(row, 'the finished list rendered nothing');
  assert.equal(row.querySelector('.cerendedname')?.textContent, 'The Elm Row lease',
    `the finished row reads ${JSON.stringify(row.textContent)} — a state and a date, which is `
    + 'five identical lines to anyone who has completed five proceedings. The name travelled with '
    + 'the folder ADR-012 moves and is the only text on the row a person recognises');
});

test('a finished ceremony that was never named renders no empty slot', async () => {
  await render({
    ceremonies: [],
    ended: [{ ceremony: ID, state: 'completed', observed_at: '2026-09-01T10:00:00Z' }],
    primary: true,
  });
  assert.equal(doc.querySelector('.cerendedname'), null,
    'an unnamed finished ceremony renders an empty name element, which pushes the date out of '
    + 'line with every other row in the column');
});

File: internal/server/ceremonyname_test.go
package server

import (
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"testing"
	"time"

	"nib/internal/ceremony"
)

// Naming a proceeding on this machine.
//
// # The defect these drive
//
// The panel headed every card with its Intent, and the Intent is the recital — a clause meant to be
// signed, not a handle to scan a list by. A user with three live proceedings over three leases read
// three cards that opened *"We agree to…"* and told them apart by squinting. Worse, a party before
// their hop has no record at all, so their card had no heading of its own to read: it fell to
// *"A ceremony on this machine"*, which is what every one of them says.

// nameCeremony posts a label and returns the code and body.
func nameCeremony(t *testing.T, c *http.Client, csrf, base, id, name string) (int, string) {
	t.Helper()
	return postForCode(t, c, csrf, base+"/api/ceremony/name",
		ceremonyNameRequest{Ceremony: id, Name: name})
}

// acceptedCeremony gets this machine to the state a named-before-anything-arrives party is in:
// an invitation accepted, a `me` marker on disk, and no record.
func acceptedCeremony(t *testing.T, c *http.Client, csrf, base string) string {
	t.Helper()
	me := myFingerprint(t, c, base)
	invitation, _ := inviteFor(t, me)
	code, body := postForCode(t, c, csrf, base+"/api/ceremony/accept",
		acceptRequest{Invitation: invitation})
	if code != http.StatusOK {
		t.Fatalf("setup: accept returned %d: %s", code, body)
	}
	var ar acceptResponse
	if err := json.Unmarshal([]byte(body), &ar); err != nil {
		t.Fatal(err)
	}
	return ar.Ceremony
}

// TestAPartyBeforeTheirHopCanNameTheirCeremonyAndTheListingCarriesIt is the whole point of the
// route, driven at the party who most needs it.
//
// **The pre-hop party is the case, not a corner of it.** They hold no record — `ReadStored`
// answers `LoadAbsent` with a `me` marker — so the server must accept a name for a ceremony it
// cannot describe, and the listing must carry that name back out of a directory holding nothing
// but two small files. A route gated on `state == ok` would pass a happy-path test and refuse
// exactly the party whose card is otherwise blank.
func TestAPartyBeforeTheirHopCanNameTheirCeremonyAndTheListingCarriesIt(t *testing.T) {
	ts, _ := startServerWith(t)
	c, csrf := authedClient(t, ts)
	id := acceptedCeremony(t, c, csrf, ts.URL)

	// SETUP: this really is the pre-hop shape. Without it the test could be passing over a
	// ceremony with a record, which is the easy case and not the one that was broken.
	if st := ceremony.ReadStored(defaultOutputDir(), id, time.Now()); st.State != ceremony.LoadAbsent || !st.Joined {
		t.Fatalf("setup: wanted an accepted, pre-hop ceremony (absent+joined); got state=%v joined=%v",
			st.State, st.Joined)
	}

	if code, body := nameCeremony(t, c, csrf, ts.URL, id, "The Elm Row lease"); code != http.StatusOK {
		t.Fatalf("naming an accepted ceremony returned %d: %s — the party with no record yet is "+
			"the one whose card says nothing but \"A ceremony on this machine\"", code, body)
	}

	got := listedCeremony(t, ts.URL, id)
	if got.Name != "The Elm Row lease" {
		t.Errorf("the listing carried name %q, want %q — a name the panel cannot read is a file "+
			"on disk and nothing else", got.Name, "The Elm Row lease")
	}
}

// TestANameIsRefusedForACeremonyThisMachineHasNeverHeardOf keeps the route from minting
// directories under ~/nib for ids nobody sent.
func TestANameIsRefusedForACeremonyThisMachineHasNeverHeardOf(t *testing.T) {
	ts, _ := startServerWith(t)
	c, csrf := authedClient(t, ts)
	unknown := strings.Repeat("ab", 16)

	code, _ := nameCeremony(t, c, csrf, ts.URL, unknown, "not mine")
	if code != http.StatusNotFound {
		t.Errorf("naming an unknown ceremony returned %d, want 404 — a label on nothing is a "+
			"directory the close-out will never visit", code)
	}
	// And nothing was created on the way to refusing — the path `WriteName` would have used, so
	// this fails if the refusal is removed rather than passing on a path nothing writes.
	dir, err := ceremony.MirrorDir(defaultOutputDir(), unknown)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(dir); !os.IsNotExist(err) {
		t.Errorf("the refused name left %s behind (%v)", dir, err)
	}
}

// TestClearingANameRemovesItRatherThanStoringABlank — "never named" and "cleared" are one state.
func TestClearingANameRemovesItRatherThanStoringABlank(t *testing.T) {
	ts, _ := startServerWith(t)
	c, csrf := authedClient(t, ts)
	id := acceptedCeremony(t, c, csrf, ts.URL)

	if code, body := nameCeremony(t, c, csrf, ts.URL, id, "The Elm Row lease"); code != http.StatusOK {
		t.Fatalf("setup: naming returned %d: %s", code, body)
	}
	if code, body := nameCeremony(t, c, csrf, ts.URL, id, "   "); code != http.StatusOK {
		t.Fatalf("clearing returned %d: %s", code, body)
	}
	if got := listedCeremony(t, ts.URL, id).Name; got != "" {
		t.Errorf("clearing left the name %q", got)
	}
	// The file, not just the field: a blank file would make the panel render an empty heading
	// where it should fall back to the recital.
	dir, err := ceremony.MirrorDir(defaultOutputDir(), id)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(filepath.Join(dir, "name")); !os.IsNotExist(err) {
		t.Errorf("clearing the name left the file behind (%v) — \"never named\" and \"cleared\" "+
			"are supposed to be the same state on disk", err)
	}
}

// TestAnOverlongNameIsCappedAndTheAnswerIsWhatWasSTORED.
//
// **The echo is the assertion.** A route that trimmed on disk and echoed the request would tell
// the client a name it does not have, and the client writes that straight back into the field —
// so the next save would be of a name the user never typed.
func TestAnOverlongNameIsCappedAndTheAnswerIsWhatWasSTORED(t *testing.T) {
	ts, _ := startServerWith(t)
	c, csrf := authedClient(t, ts)
	id := acceptedCeremony(t, c, csrf, ts.URL)

	long := strings.Repeat("é", ceremony.MaxCeremonyNameLen+40)
	code, body := nameCeremony(t, c, csrf, ts.URL, id, long)
	if code != http.StatusOK {
		t.Fatalf("an over-long name returned %d: %s — it is capped, not refused", code, body)
	}
	var out struct {
		Name string `json:"name"`
	}
	if err := json.Unmarshal([]byte(body), &out); err != nil {
		t.Fatal(err)
	}
	if n := len([]rune(out.Name)); n != ceremony.MaxCeremonyNameLen {
		t.Errorf("the answer carried %d runes, want %d — the client refills its field from this, "+
			"so an echo of the REQUEST hands back a name the machine does not hold", n,
			ceremony.MaxCeremonyNameLen)
	}
	if stored := listedCeremony(t, ts.URL, id).Name; stored != out.Name {
		t.Errorf("the route answered %q and stored %q", out.Name, stored)
	}
}

// TestTheNameFieldInTheClientMatchesTheServersCap holds one number in two languages together.
//
// A `maxlength` larger than the server's cap silently truncates what the user typed; smaller, and
// the field refuses characters the server would have kept. Neither shows up as a failure anywhere.
func TestTheNameFieldInTheClientMatchesTheServersCap(t *testing.T) {
	b, err := os.ReadFile("../../web/app.js")
	if err != nil {
		t.Fatal(err)
	}
	m := regexp.MustCompile(`field\.maxLength = (\d+);`).FindSubmatch(b)
	if m == nil {
		t.Fatal("web/app.js sets no `field.maxLength` on the ceremony name field — either the " +
			"control lost its cap or it was renamed and this guard now checks nothing")
	}
	if want := strconv.Itoa(ceremony.MaxCeremonyNameLen); string(m[1]) != want {
		t.Errorf("web/app.js caps the name field at %s and ceremony.MaxCeremonyNameLen is %s",
			m[1], want)
	}
}

// listedCeremony reads one entry out of the panel's own listing, which is the only path the client
// ever sees a name by.
func listedCeremony(t *testing.T, base, id string) ceremony.Stored {
	t.Helper()
	for _, s := range getCeremonies(t, base).Ceremonies {
		if s.ID == id {
			return s
		}
	}
	t.Fatalf("the listing has no ceremony %s", id)
	return ceremony.Stored{}
}

// TestAFinishedCeremonyKeepsItsNameAndTheReceiptNeverStoresIt.
//
// **Two halves of one rule and they pull against each other.** The name has to survive the
// close-out, because a finished row otherwise carries a state and a date and nothing a person
// recognises — five completed proceedings read as five identical lines. And it must NOT be copied
// into `receipt.json`, which is write-once by design: a snapshot there would disagree with the
// `name` file the moment either changed, and `WriteReceipt`'s conflict rule would then be
// comparing a state against a stale record.
func TestAFinishedCeremonyKeepsItsNameAndTheReceiptNeverStoresIt(t *testing.T) {
	ts, _ := startServerWith(t)
	c, csrf := authedClient(t, ts)
	id := acceptedCeremony(t, c, csrf, ts.URL)
	root := defaultOutputDir()

	if code, body := nameCeremony(t, c, csrf, ts.URL, id, "The Elm Row lease"); code != http.StatusOK {
		t.Fatalf("setup: naming returned %d: %s", code, body)
	}
	// ADR-012's close-out: the folder MOVES, and the receipt lands beside what it moved.
	if err := ceremony.CloseOutMirror(root, id); err != nil {
		t.Fatalf("setup: close-out: %v", err)
	}
	if err := ceremony.WriteReceipt(root, id, ceremony.Receipt{
		Ceremony: id, State: ceremony.StateLeft, ObservedAt: time.Now(),
	}); err != nil {
		t.Fatalf("setup: receipt: %v", err)
	}

	ended, err := ceremony.ListEnded(root)
	if err != nil {
		t.Fatal(err)
	}
	if len(ended) != 1 {
		t.Fatalf("setup: %d ended ceremonies, want 1", len(ended))
	}
	if ended[0].Name != "The Elm Row lease" {
		t.Errorf("the finished row carried name %q, want %q — the name travels with the folder "+
			"ADR-012 moves, and it is the only text on that row a person would recognise",
			ended[0].Name, "The Elm Row lease")
	}

	dir, err := ceremony.EndedDir(root, id)
	if err != nil {
		t.Fatal(err)
	}
	raw, err := os.ReadFile(filepath.Join(dir, "receipt.json"))
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(string(raw), "Elm Row") {
		t.Errorf("receipt.json stored the name:\n%s\nThe receipt is write-once and the name is "+
			"editable, so a copy here is a snapshot that disagrees with the `name` file as soon "+
			"as either moves", raw)
	}
}
